xarray/core/dataset.py

```diff
@@ -1955,7 +1955,35 @@ def fillna(self, value):
         -------
         Dataset
         """
-        out = self._fillna(value)
+        if utils.is_dict_like(value):
+            value_keys = value.data_vars.keys() if isinstance(value, Dataset)\
+                else value.keys()
+            if not set(value_keys) <= set(self.data_vars.keys()):
```

Comment:
I think we will need to separate the `join` argument to `apply_ufunc` into two parts:

1. The `join` used for [aligning indexes](https://github.com/pydata/xarray/blob/4d36fe2e8a1f5504d776c2d69895bc4d5a9cd31a/xarray/core/computation.py#L340).
2. The `join` used for [aligning data variables between datasets](https://github.com/pydata/xarray/blob/4d36fe2e8a1f5504d776c2d69895bc4d5a9cd31a/xarray/core/computation.py#L347).

The later should probably be renamed to something like `data_vars_join`